torchdata/dataloader2/reading_service.py

```diff
@@ -381,35 +388,44 @@ def _pause(self):
                 continue
             if hasattr(dp, "pause") and callable(dp.pause):
                 dp.pause()
-        if self.num_workers > 0:
-            self._worker_consumer_datapipe.request_pause()  # type: ignore[union-attr]
-        else:
-            raise RuntimeError(
-                "If you would like to use `pause` with `MultiProcessingReadingService`, "
-                "please use more than 0 worker."
-            )
+        self._worker_consumer_datapipe.request_pause(pause_fn)  # type: ignore[union-attr]
+        return None
 
-    def _resume(self):
+    def _resume(
+        self, resume_fn: Optional[Callable[[DataPipe], DataPipe]] = None
+    ) -> Optional[Callable[[DataPipe], DataPipe]]:
         """
         Resumes DataPipes' activities. This is required to be called after `_pause` before
         the DataLoader can keep yielding elements.
         """
         if self.num_workers > 0:
-            self._worker_consumer_datapipe.request_resume()  # type: ignore[union-attr]
+            self._worker_consumer_datapipe.request_resume(resume_fn)  # type: ignore[union-attr]
         else:
             raise RuntimeError(
                 "If you would like to use `resume` with `MultiProcessingReadingService`, "
                 "please use more than 0 worker."
             )
-        if self.main_prefetch_cnt > 0 and self.num_workers > 0:
-            self._main_prefetch_datapipe.resume()  # type: ignore[union-attr]
+        assert self._end_datapipe is not None
+        dp_list = list_dps(traverse_dps(self._end_datapipe))
+        for dp in dp_list[::-1]:
+            # TODO: Combine QueueWrapper and _IterateQueueDataPipes,
+            #       and attach resume method. Then, no need to call
+            #       self._worker_consumer_datapipe.request_resume()
+            if isinstance(dp, communication.iter.QueueWrapper):
+                continue
+            if hasattr(dp, "resume") and callable(dp.resume):
+                dp.resume()
+        return None
```

Comment:
What DataPipes do we need to `resume`/`pause` here aside from the `prefetch` for main loop? Dispatching process (seems to be handled elsewhere)? `FullSync` (unlikely)?

optional nit: Maybe add a comment before `assert` to indicate the purpose?

Reply:
> With this PR, is pause/resume full functional with dispatching process? Or is something still missing?

Yes, it is. I might add one more test to validate it.

> What DataPipes do we need to `resume`/`pause` here aside from the `prefetch` for main loop? Dispatching process (seems to be handled elsewhere)? `FullSync` (unlikely)?

Two possible DataPipes. One is fullsync and prefetch, another case is some user-defined non-replicable datapipe with `resume`/`pause` implemented (in the main process)

> optional nit: Maybe add a comment before `assert` to indicate the purpose?

Sure, I can do that.